README.md
wavnorm
=======

wavnorm - automatically maximizes the level of a .wav file, destructively, without using a temp file.

nplay - simple text mode .wav file player.

nrecord - simple text mode .wav file recorder, with level and clip monitor.

IMPORTANT NOTE: These were written and tested on 32 bit Intel architecture machines (i386, i686) - they do not appear to work due to the struct packing on other architectures, i.e. 32 bit ARM (R-Pi) or 64 bit x86_64

wavnorm, nplay, nrecord

These are some simple but handy audio tools that I wrote to fill a need
where I couldn't find similar tools that would do the job. All programs are
presently limited to 16bit stereo wav files at 44100 Hz, as they are
intended for preparing music for making demo CDs and/or mp3 files.

nplay and nrecord presume OSS sound drivers, or an emulation (i.e. /dev/dspX audio devices).

THESE PROGRAMS HAVE NO WARRANTY! They work fine for me but if they don't for
you you are welcome to read the code and send me a patch! 

They are placed under the GPL (see the file COPYING for details) and made
available "as is" in case anyone else finds it handy as well.

Compilation: wavnorm should compile on anything. nplay and nrecord need the
newt user interface library that is shipped with RedHat Linux. This is GPL
code and should be readily available for other systems.

wavnorm - automatically maximize the level of a .wav file.

This is a simple command line, destructive "normalizer" for 16 bit, stereo
wav files. By destructive I mean it works on the ORIGINAL file specified on
the command line, it does NOT make a copy. If you want more flexible
options, the sox package can do all this and more, but with a bit more
fiddling with command line options.

What it actually does is find the highest level in a .wav file, and then
scale the entire .wav file so that the loudest point is as loud as possible
without clipping. It does this "in place" on the original .wav file, so
space consuming temp files aren't created. This is handy when creating demo
compilation CDs of home recorded material from different sources, which can
vary widely in levels. wavnorm can impose SOME degree of consistency.

I make no claims about the audio quality - it sounds fine to me, if that is
your concern hire a mastering house to do it properly.

Usage: note that the program only actually alters anything if you specify
the "-d" (for DESTRUCTIVE) command line option.

The "-t" option can be used to specify a threshold under which it wont make
a change.. for example if the calculated ratio is 1.03 scaling up the .wav
wont make it that much louder, but probably will alter the sound a tiny bit,
so we may as well skip scaling up the level.

If wavnorm finds the length of the file as indicated by the .wav header is
not what it should be, it will update the length as well if run in
destructive mode.

wavnorm runs very quickly but is quite intensive in its use of system
resources. It is not a good idea to record audio with nrecord while running
wavnorm.

nplay - play a wav file

This is a very simple .wav player. It uses a newt/ncurses interface and will
play all the .wav files specified on the command line. It has a simple bar
graph to indicate how far through the file it is, plus 2 VU meter graphs one
to indicate the current max level in the current interval, the other to
indicate the total peak level hit so far. The interval in which the screen
is updated is set by the -i option, but defaults to 0.1 (10 times per
second).

nrecord - record a wav file, while monitoring the levels

nrecord records 16 bit stereo audio at 44100 Hz from the /dev/dsp device (or
another device if specified by the -d option) and saves it as a .wav file,
until enter is pressed. It uses a newt/ncurses interface to give a count of
elapsed time in seconds, and 2 VU meters, one indicating the current level,
and one indicated the highest level attained so far. If the level has hit
maximum, the word CLIPPED is displayed on the screen. The advantage of this
program is it's simplicity, as it allows simple stereo wav recording with
level monitoring without fiddling around with resource hungry multitrack
software. Run aumix or a similar mixer program in another xterm or console,
adjust the levels so there is enough headroom that clipping wont occur and
record.

LINUX KERNEL ISSUES

When using wavnorm with the Linux 2.2.17 kernel I get strange errors and
occasional lockups. This did not occur with the 2.2.18 kernel, which has 
some fixes for the VM stuff. wavnorm relies heavily on mmap(), so presumably
it was hitting the VM bugs that were fixed in 2.2.18.

CURRENT STATUS

Although I could add lots of features to this program, there are other
bits of software such as ecasound that do a range of command line audio
tasks very well. This program is intended to do just one thing but do it
quickly, and without needing temporary files. 
I have been very busy with other stuff (music, software, life) in the year
since 0.4, and have not had much time to add to this project, although it
is essentially feature complete. Thanks to all who have sent in patches,
suggestions and encouragment, I have tried to take your ideas on board
where possible.


duncan@zog.net.au 19/07/1999 - 14/07/2001

http://www.linuxbandwagon.com/
http://www.zog.net.au/

addendum, June 2014:
I've checked these into github "as is" to make the code available. Right now I havent checked that it builds or functions on a modern distro, but there is no reason it shouldn't.

Building on a Ubuntu/Debian system - these will need the libnewt-dev package installed, i.e.
sudo apt-get install libnewt-dev



Changelog

http://www.linuxbandwagon.com/wavnorm/

Changelog:

0.1 - created from frags of code from my aborted sound editor.

0.2 - changed name to wavnorm, added -r option, cleaned up code,
      added GPL and README

0.3 - Cleaned up error checking, now handles more than one wav file on
      the command line, made the output of the program saner.

0.4 - Wrote nrecord and nplay, modularised the code further, added the
      fix wav length function, etc. Put this in a seperate file and
      put everything under CVS.

0.5 - Added updates and manpages from Jorgen Grahn <jgrahn@algonet.se>
    - Lost CVS tree due to various software changes I won't go into, so
      took out the CVS stuff for now.
    - Added improvements to wavnorm from Frederik Arbeiter 
      <frederik.arbeiter@gmx.net>, see
      http://www.uni-karlsruhe.de/~ualy/software/wavlevel/wavlevel.html
      for wavlevel, his handy package!
    - Also added suggestions by Chad Hurwitz <churritz@king.cts.com>,
      regarding allowing some clipping to occur in a controlled manner.
      Part of this also includes a simple display of where the peaks
      occur in the file, handy for choosing compression or choosing
      how many clips to include. Still unsure of the usefulness of this
      code however. 
    - Added scaling function as in the patch sent in by Attila Bogar,
      <boat@elte.hu>
    - Added a little % progress thingy which is better than how it was
      before.

0.5github - No code change but checked into GitHub June 2014
      


nplay.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <unistd.h>
#include <sys/mman.h>
#include <sys/ioctl.h>
#include <linux/soundcard.h>
#include <newt.h>
#include <limits.h>

#include "wavfuncs.h"

/****

nplay - simple wav player

duncan@zog.net.au

$Id: nplay.c,v 1.3 1999/07/28 13:18:06 duncan Exp $

*****/

/** Needed to compile on RH 4.X for some reason **/

#ifndef MAP_FAILED
#define MAP_FAILED -1
#endif

double interval_s = 0.1;  /* default interval in seconds */

int play_buffer(int dspfd, short *start, short *end);
int open_dsp(WAVE_HEADER *wav_info);

void usage(char *progname);
void play_wav_file(char *filename);

int main(int argc,char *argv[])
{
   int c;

   while ((c = getopt(argc, argv, "i:")) != EOF)
   {
      switch(c)
      {
         case 'i':
            interval_s = atof(optarg);
            break;
         default:
            usage(argv[0]);
            exit(1);
         }
   }

   if ((optind + 1) > argc)
   {
      usage(argv[0]);
      exit(1);
   }

   newtInit();

   for (c = optind; c < argc; c++)
   {
      play_wav_file(argv[c]);
   }

   newtFinished();         

   return(0);
}


void play_wav_file(char *filename)
{
   char *inwavbuf;
   short *current;
   short *audioend;
   short *audio;
   WAVE_HEADER *wav_info;
   int wavfd;
   int dspfd;
   struct stat input_fstat;
   size_t interval;

   newtComponent vu_1sec;
   newtComponent vu_total;
   newtComponent wav_length;
   newtComponent label_length;
   newtComponent mainwaveform;
   newtComponent label_1sec;
   newtComponent label_total;
   newtComponent rf_result;

   char labelstr_1sec[10] = "0";
   char labelstr_total[10] = "0";
   char labelstr_length[10] = "";
   short one_sec_max, total_max;

   wavfd = open(filename,O_RDONLY,0600);
   if (wavfd == -1)
   {
      printf("Error: open() %s\n",strerror(errno));
      exit(1);
   }

   if (fstat(wavfd,&input_fstat) != 0)
   {
      printf("Error: fstat() %s\n",strerror(errno));
      return;
   }

   if (input_fstat.st_size < sizeof(WAVE_HEADER))
   {
      printf("File is not large enough to hold a .wav file header even!\n");
      return;
   }


   inwavbuf = mmap(NULL,input_fstat.st_size,PROT_READ,MAP_SHARED,wavfd,0);
   if (inwavbuf == MAP_FAILED)
   {
      printf("Error: mmap() %s\n",strerror(errno));
      exit(1);
   }


   audio = (short *)validate_wav_header(inwavbuf,0);
   current = audio;

   if (current == NULL)
   {
      printf("This program didn't like the wav file\n");
      exit(1);
   }

   wav_info = (WAVE_HEADER *)inwavbuf;
   audioend =  (short *)((char *)audio + wav_info->nDataBytes);

   dspfd = open_dsp(wav_info);

   newtCls();

   newtDrawRootText(0, 0, filename);
   mainwaveform = newtForm(NULL, NULL,  NEWT_FLAG_NOF12);
   vu_1sec    =  newtScale(9,5,68,(long long)(SHRT_MAX));
   label_1sec =  newtLabel(1,5,labelstr_1sec);
   wav_length  =  newtScale(9,3,68,audioend - audio);
   label_length = newtLabel(1,3,labelstr_length);
   vu_total =  newtScale(9,8,68,(long long)(SHRT_MAX));
   label_total =  newtLabel(1,8,labelstr_total);
   
   newtFormAddComponent(mainwaveform,vu_1sec);
   newtFormAddComponent(mainwaveform,vu_total);
   newtFormAddComponent(mainwaveform,label_1sec);
   newtFormAddComponent(mainwaveform,label_total);

   one_sec_max = 0;
   total_max = 0;
   newtFormWatchFd(mainwaveform,dspfd,NEWT_FD_WRITE); 
   newtFormAddHotKey(mainwaveform,NEWT_KEY_ENTER);
   newtPushHelpLine("Hit Enter to end playing");
   newtCenteredWindow(78,10,"now playing .wav file");

   newtRefresh();

   /* presently every second */
   interval = (size_t )((double )wav_info->nSamplesPerSec * interval_s * 2);

   while ((current) < audioend)
   {
      short *endcurrent = current + interval;

      if (endcurrent > audioend)
      {
         endcurrent = audioend;
      }
         
      one_sec_max = get_peak_value(current,endcurrent);
      newtScaleSet(vu_1sec,one_sec_max);
      sprintf(labelstr_1sec,"%1.6f",((float )one_sec_max/ (float )SHRT_MAX));
      newtLabelSetText(label_1sec,labelstr_1sec);
      newtScaleSet(wav_length,current - audio);
      sprintf(labelstr_length,"%4.2f",
         ((double )(current - audio) / 88200));
      newtLabelSetText(label_length,labelstr_length);
      if (one_sec_max > total_max)
      {
         total_max = one_sec_max;
         sprintf(labelstr_total,"%1.6f",((float )total_max/ (float )SHRT_MAX));
         newtLabelSetText(label_total,labelstr_total);
         newtScaleSet(vu_total,total_max);
      }
      rf_result = newtRunForm(mainwaveform);
      if (play_buffer(dspfd,current,endcurrent) == -1)
      {
         current = audioend;
      }
      
      current = endcurrent;
      if (rf_result == NULL)
         current = audioend;

      newtRefresh();
   }

   newtFormDestroy(mainwaveform);

   munmap(inwavbuf,input_fstat.st_size);
   close(wavfd);               
   close(dspfd);
      
   return;
}

/**

Opens up /dev/dsp to play audio data in a way that is
compatible with the given .wav header.

**/

int open_dsp(WAVE_HEADER *wav_info)
{
   int dspfd;
   int status;
   int arg;

   /** open the sound device and set it up **/
   dspfd = open("/dev/dsp",O_WRONLY);
   if (dspfd < 0) 
   {
      perror("open of /dev/dsp failed");
      return(-1);
   }
   arg = wav_info->wBitsPerSample;
   status = ioctl(dspfd, SOUND_PCM_WRITE_BITS, &arg);

   if (status == -1)
   {
      perror("SOUND_PCM_WRITE_BITS ioctl failed");
      return(-1);
   }
   if (arg != wav_info->wBitsPerSample)
   {
      perror("unable to set sample size");
      return(-1);
   }

   arg = wav_info->nChannels;  /* mono or stereo */
   status = ioctl(dspfd, SOUND_PCM_WRITE_CHANNELS, &arg);
   if (status == -1)
   {
      perror("SOUND_PCM_WRITE_CHANNELS ioctl failed");
      return(-1);
   }
   if (arg != wav_info->nChannels)
   {
      perror("unable to set number of channels");
      return(-1);
   }

   arg = wav_info->nSamplesPerSec;      /* sampling rate */
   status = ioctl(dspfd, SOUND_PCM_WRITE_RATE, &arg);
   if (status == -1)
   {
      perror("SOUND_PCM_WRITE_WRITE ioctl failed");
      return(-1);
   }
   return(dspfd);
}

int play_buffer(int dspfd, short *start, short *end)
{
   int status;

   status = write(dspfd,(char *)start,(char *)end - (char *)start); 
          /* play it back */

   if (status != (char *)end - (char *)start)
   {
      perror("wrote wrong number of bytes");
      return(-1);
   }
   
   return(0);
}

void usage(char *progname)
{
   printf("%s [-i interval] file.wav [file.wav ...]\n",
      progname);
   printf("   -i interval in seconds to update VU & screen, default 1.0\n");
}



wavnorm.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/mman.h>
#include <math.h>
#include <limits.h>

#include "wavfuncs.h"

/**

wavnorm - a 16 bit stereo .wav file normaliser

duncan@zog.net.au 19/7/1999

$Id: wavnorm.c,v 1.5 1999/07/28 13:18:06 duncan Exp $

**/



/** Needed to compile on RH 4.X for some reason **/

#ifndef MAP_FAILED
#define MAP_FAILED -1
#endif

int wavfd;

/** command line options - stored as globals **/
int verbose = 0;             /* -v */ 
int histogram = 0;           /* -h */
int destructive = 0;         /* -d */
double threshold = -1.0;     /* -t */
double ratio = 1.0;          /* -r */
double scale = 0.0;          /* -s */
long overs_allowed = 0;      /* -o */

/** a big ugly array to store the map of peaks **/
long levels[SHRT_MAX];

void usage(char *progname);
void normalise_wav_file(char *filename);

int main(int argc,char *argv[])
{
   int c;

   while ((c = getopt(argc, argv, "vhdt:r:o:s:")) != EOF) 
   {
      switch(c) 
      {
         case 'v':
            verbose = 1;
            printf("%s version %s duncan@zog.net.au 1999\n",argv[0],VERSION);
            break;
         case 'h':
            /* i know this isnt really a histogram, but it gives that
               kind of info... */
            histogram = 1;
         case 'd':
            destructive = 1;
            break;
         case 't':
            threshold = atof(optarg);
            if (verbose)
               printf("Using threshold of ratio of %f\n",threshold);
            break;
         case 'r':
            ratio = atof(optarg);
            break;
         case 'o':
            overs_allowed = atol(optarg);
            break;
         case 's':
            scale = atof(optarg);
            break;

         default:
            usage(argv[0]);
            exit(1);
         }
   }

   if ((optind + 1) > argc)
   {
      usage(argv[0]);
      exit(1);
   }

   for (c = optind; c < argc; c++)
   {
      normalise_wav_file(argv[c]);
   }

   return(0);
}


void normalise_wav_file(char *filename)
{      
   char *inwavbuf;
   short *current;
   short *audio;
   short *audioend;
   WAVE_HEADER *wav_info;
   int wavfd;
   struct stat input_fstat;
   size_t increment;
   int loudest = 0;
   double loudestd;
   double loudr;
   int i;

   printf("processing %s\n",filename);

   wavfd = open(filename,O_RDWR,0600);
   if (wavfd == -1)
   {
      printf("Error: open() %s\n",strerror(errno));
      return;
   }

   if (fstat(wavfd,&input_fstat) != 0)
   {
      printf("Error: fstat() %s\n",strerror(errno));
      if (close(wavfd) != 0)
      {
         printf("Error: close() %s\n",strerror(errno));
         exit(1);
      }
      return;
   }

   if (input_fstat.st_size < sizeof(WAVE_HEADER))
   {
      printf("Error: %s not large enough to hold a .wav file header!\n",
         filename);
      return;
   }

   inwavbuf = mmap(NULL,input_fstat.st_size,
      PROT_READ | PROT_WRITE,MAP_SHARED,wavfd,0);
   if (inwavbuf == MAP_FAILED)
   {
      printf("Error: mmap() %s\n",strerror(errno));
      return;
   }

   audio = (short *)validate_wav_header(inwavbuf,verbose);

   if (audio == NULL)
   {
      printf("Error: %s not in suitable .wav format for this program\n",
         filename);
      return;
   }

   wav_info = (WAVE_HEADER *)inwavbuf;

   increment = wav_info->wBitsPerSample / 8;

   if (wav_info->nDataBytes != input_fstat.st_size - sizeof(WAVE_HEADER))
   {
      printf("   Warning: file length does not match length in header.\n");

      if (destructive)
      {
         printf("   --> fixing length to correct value\n");
         wav_info->nDataBytes = input_fstat.st_size - sizeof(WAVE_HEADER);
      }
   }

   audioend = (short *)((char *)audio + wav_info->nDataBytes);

   if (scale == 0.0)
   {
      current = audio;

      if (verbose)
         printf("   scanning levels in wav file:\n");

      memset(levels,0,sizeof(levels));

      /** research phase - run through the .wav and gather some info **/

      while (current < audioend)
      {
         levels[abs(*current)]++;
         if ((verbose) && ! ((long )current % 21))
            progressbar_peak(audio,current,audioend,loudest);
         if (abs(*current) > loudest)
         {
            loudest = abs(*current);
         }
         current++;
      }
      if (verbose)
      {
         progressbar_peak(audio,current,audioend,loudest);
         printf("\n");
      }
      current--;

      /** development phase - use our info to figure out what we're gonna do
       *  then put out some info as well
       */

      if (1) /* but we might not always do this bit? */
      {
         long total = 0;
         long last  = 0;
         int increment = 1;
         int nextstep = 1;

         if (histogram)
            printf("level\t\t%%of points\t\t# of points\n");

         for (i = SHRT_MAX; i > 0; i--)
         {
            if (levels[i] > 0)
            {
               last = total;
	       total = levels[i] + total;
               if ((overs_allowed <= total) && (overs_allowed > last))
               {
                  loudest = i;
                  if (histogram)
                     printf("%3.4f\t\t%3.4f\t\t\t%6li <-- found %ld overs\n",
                        (double )i/(double )SHRT_MAX,
                        (double )total / (double )(audioend - audio) * 100,
                        total,
                        overs_allowed);
               }
            }
            if (total >= nextstep)
            {
               if (histogram)
                  printf("%3.4f\t\t%3.4f\t\t\t%6li\n",
                     (double )i/(double )SHRT_MAX,
                     (double )total / (double )(audioend - audio) * 100,
                     total);
               nextstep += (increment * 3);
               if (total >= increment * 10)
               {
                  nextstep = total;
                  increment = increment * 10;
               }
            }
         }
      }
      loudestd = fabs((double )loudest);
      loudr =  SHRT_MAX / loudestd * ratio;
      if (verbose)
         printf("    with %ld digital overs allowed\n",overs_allowed);
   }
   else
   {
      current = audioend;
      threshold = 0;
      loudr = scale;
   }

   if (verbose)
   {
      printf("    scaling to ratio %f of max\n",ratio);
      printf("    normalise ratio is %f\n",loudr);
   }

   /** production - lets mess with the file! **/

   if (destructive)
   {
      if (loudr > threshold)
      {
         if (verbose)
            printf("    normalising with ratio %f now...\n",loudr);

         /** we go backwards through the audio now, as presumably the
             last bytes we read are more likely to be buffered, so
             hopefully this is a bit faster? **/

  
         while (current > audio)
         {
            if ((verbose) && ! ((long )current % 21))
            {
               progressbar(audio,current,audioend);
            }

            *current = (int )((double )*current * loudr);
            current--;
         }
         if (verbose)
         {
            printf("\rdone: \t %2.2f%%     ",100.000);
            printf("\n");
         }
      }
      else
      {
         if (verbose)
            printf("   threshold not hit, file not altered\n");
      }
   }

   munmap(inwavbuf,input_fstat.st_size);

   if (close(wavfd) != 0)
   {
      printf("Error: close() %s\n",strerror(errno));
      exit(1);
   }
      
}

void usage(char *progname)
{
   printf("%s: version %s\n",progname,VERSION);
   printf("%s <options> file.wav [file2.wav ...]\n",
      progname);
   printf("   where <options> are one or more of:\n");
   printf("   -v verbose, and version info\n");
   printf("   -d destructive (actually alter the .wav!)\n");
   printf("   -t <threshold> threshhold ratio under which dont alter\n");
   printf("   -r <ratio>     ratio of max possible to scale (default 1.0)\n");
   printf("   -s <scale>     force scaling to this ratio, dont look for a max\n");
   printf("   -o <overs>     allow <overs> number of peaks to exceed max when looking for highest\n");
   printf("   -h             print histogram type diagram of where the highest points are in the file\n");

}


wavfuncs.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/mman.h>
#include <math.h>

#include "wavfuncs.h"


/**

$Id: wavfuncs.c,v 1.2 1999/07/28 12:49:48 duncan Exp $

duncan@zog.net.au 19/7/1999


**/



/**

uses info from 
http://www-ccrma.stanford.edu/CCRMA/Courses/422/projects/WaveFormat/

returns a pointer to the start of the audio data, or NULL if the
wav header isnt what we want.

**/

char *validate_wav_header(char *header, int verbose)
{
   WAVE_HEADER *wheader = (WAVE_HEADER *)header;

   /* first 4 bytes should be RIFF */

   if (strncmp(wheader->RiffID,"RIFF",4) != 0)
   {
      if (verbose)
         printf(".wav header dont have RIFF\n");
      return(NULL);
   }

   /* if (verbose) printf("RiffSize is %ld\n",wheader->RiffSize); */

   /* next 4 bytes should be WAVE */
   if (strncmp(wheader->WaveID,"WAVE",4) != 0)
   {
      if (verbose) printf("header dont have WAVE\n");
      return(NULL);
   }          

   /* next 4 bytes should be "fmt " */
   if (strncmp(wheader->FmtID,"fmt ",4) != 0)
   {
      if (verbose)
         printf("header dont have fmt\n");
      return(NULL);
   }           

   if (wheader->FmtSize != 16)
   {
      if (verbose)
         printf("this program only deals with 16 bit audio data\n");
      return(NULL);
   }

   if (wheader->wFormatTag != 1)
   {
      if (verbose)
         printf("This wav is not PCM, we cant deal with it\n");
      return(NULL);
   }

   if (wheader->nChannels != 2)
   {
      if (verbose)
          printf("Presently we only deal with two channels, this has %d\n",
             wheader->nChannels);
      return(NULL);
   }        
   if (wheader->nSamplesPerSec != 44100)
   {
      if (verbose) 
         printf("Presently we only deal with 44100 hz, this has %ld\n",
            wheader->nSamplesPerSec);
      return(NULL);
   }  

   /* next 4 bytes should be "data" */
   if (strncmp(wheader->DataID,"data",4) != 0)
   {
      if (verbose) printf("header dont have data\n");
      return(NULL);
   }           

   return (header + sizeof(WAVE_HEADER));
}

short get_peak_value(short *start, short *end)
{
   short the_max = 0;

   short *current = start;

   while (current < end)
   {
      if (abs(*current) > the_max)
         the_max = abs(*current);
      current++;
   }
   return the_max;
}

void create_wav_header(WAVE_HEADER *wheader,
   int nChannels,int nSamplesPerSec,int wBitsPerSample)
{
   /* first 4 bytes should be RIFF */

   strncpy(wheader->RiffID,"RIFF",4);

   /* next 4 bytes should be WAVE */
   strncpy(wheader->WaveID,"WAVE",4);

   /* next 4 bytes should be "fmt " */
   strncpy(wheader->FmtID,"fmt ",4);

   wheader->FmtSize = 16;

   wheader->wFormatTag = 1;

   wheader->nChannels = nChannels;
   wheader->nSamplesPerSec = nSamplesPerSec;
   wheader->nAvgBytesPerSec = nSamplesPerSec * nChannels * wBitsPerSample/8;
   wheader->wBitsPerSample = wBitsPerSample;
   wheader->nBlockAlign = nChannels * wBitsPerSample/8;

   strncpy(wheader->DataID,"data",4);
}


wavfuncs.h
/*

$Id: wavfuncs.h,v 1.2 1999/07/28 12:49:48 duncan Exp $

some common functions we do on wav files in this prog...

*/

#define VERSION "0.5"


/** data structure that holds the wav head info in a form we can use **/

typedef struct
{       char            RiffID[4];
        u_long          RiffSize;
        char            WaveID[4];
        char            FmtID[4];
        u_long          FmtSize;
        u_short         wFormatTag;
        u_short         nChannels;
        u_long          nSamplesPerSec;
        u_long          nAvgBytesPerSec;
        u_short         nBlockAlign;
        u_short         wBitsPerSample;
        char            DataID[4];
        u_long          nDataBytes;
} WAVE_HEADER;

void create_wav_header(WAVE_HEADER *wav_header,
   int nChannels,int nSamplesPerSec,int wBitsPerSample);
char *validate_wav_header(char *header, int verbose);
short get_peak_value(short *start, short *end);
void progressbar_peak(short *start,short *current,short *audioend, short max);
void progressbar(short *start,short *current,short *audioend);


nrecord.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <unistd.h>
#include <sys/mman.h>
#include <sys/ioctl.h>
#include <linux/soundcard.h>
#include <newt.h>
#include <limits.h>

#include "wavfuncs.h"

/****

nrecord - simple wav recorder

duncan@zog.net.au

$Id: nrecord.c,v 1.3 1999/07/28 13:18:06 duncan Exp $

****/



/** this is for recording to CD so this is the only setting I want **/
#define SAMPLES_PER_SECOND 44100
#define BITS_PER_SAMPLE    16
#define NUMBER_OF_CHANNELS 2

#define DEFAULT_DSP_DEVNAME "/dev/dsp"

/** Needed to compile on RH 4.X for some reason **/

#ifndef MAP_FAILED
#define MAP_FAILED -1
#endif

double interval_s = 1.0;  /* interval in seconds */
char *dsp_devname = DEFAULT_DSP_DEVNAME;

int open_dsp_read(char *filename,size_t *blk_size);


void usage(char *progname);
void record_wav_file(char *filename);

int main(int argc,char *argv[])
{
   int c;

   while ((c = getopt(argc, argv, "i:d:")) != EOF)
   {
      switch(c)
      {
         case 'i':
            interval_s = atof(optarg);
            printf("warning : -i option doesnt do much right now!\n");
            break;
         case 'd':
            dsp_devname = strdup(optarg);
            break;
         default:
            usage(argv[0]);
            exit(1);
         }
   }


  /* right now we only record one wav at a time... */

   if ((optind + 1) != argc)
   {
      usage(argv[0]);
      exit(1);
   }



   for (c = optind; c < argc; c++)
   {
      record_wav_file(argv[c]);
   }


   return(0);
}


void record_wav_file(char *filename)
{
   char *wavbuf;
   int wavfd;
   int dspfd;
   size_t interval;
   int running = 1;
   size_t bytes_written = 0;

   WAVE_HEADER wav_header, *wav_header_on_disk;

   newtComponent vu_1sec;
   newtComponent vu_total;
   newtComponent label_length;
   newtComponent mainwaveform;
   newtComponent label_1sec;
   newtComponent label_total;
   newtComponent rf_result;

   char labelstr_1sec[10] = "0";
   char labelstr_total[10] = "0";
   char labelstr_length[10] = "";
   short one_sec_max, total_max;

   wavfd = open(filename,O_RDWR | O_CREAT | O_TRUNC,0600);
   if (wavfd == -1)
   {
      printf("Error: open() %s\n",strerror(errno));
      exit(1);
   }

   create_wav_header(&wav_header,
      NUMBER_OF_CHANNELS,SAMPLES_PER_SECOND,BITS_PER_SAMPLE);

   if (write(wavfd,(void *)&wav_header,sizeof(wav_header)) != sizeof(wav_header))
   {
      printf("Error: write() %s\n",strerror(errno));
      exit(1);
   }

   dspfd = open_dsp_read(dsp_devname,&interval);
   if (dspfd == -1)
   {
      printf("Error getting %s ready for recording\n",dsp_devname);
   }

   interval_s = wav_header.nAvgBytesPerSec * interval;


   if ((wavbuf = malloc(interval)) == NULL)
   {
      printf("Error: malloc() %s\n",strerror(errno));
      exit(1);
   }
   

   newtInit();
   newtCls();

   newtDrawRootText(0, 0, filename);
   mainwaveform = newtForm(NULL, NULL,  NEWT_FLAG_NOF12);
   vu_1sec    =  newtScale(9,5,68,(long long)(SHRT_MAX));
   label_1sec =  newtLabel(1,5,labelstr_1sec);
   label_length = newtLabel(1,3,labelstr_length);
   vu_total =  newtScale(9,8,68,(long long)(SHRT_MAX));
   label_total =  newtLabel(1,8,labelstr_total);
   
   newtFormAddComponent(mainwaveform,vu_1sec);
   newtFormAddComponent(mainwaveform,vu_total);
   newtFormAddComponent(mainwaveform,label_1sec);
   newtFormAddComponent(mainwaveform,label_total);

   one_sec_max = 0;
   total_max = 0;
   newtFormWatchFd(mainwaveform,dspfd,NEWT_FD_READ); 
   newtFormAddHotKey(mainwaveform,NEWT_KEY_ENTER);
   newtPushHelpLine("Hit Enter to end recording");
   newtCenteredWindow(78,10,"now recording .wav file");

   newtRefresh();

   /* presently every second */

   while (running)
   {
      if (read(dspfd,wavbuf,interval) != interval)
      {
         newtFinished();         
         printf("read(dsp) %s",strerror(errno));
         exit(1);
      }
      if (write(wavfd,wavbuf,interval) != interval)
      {
         newtFinished();
         printf("write(%s) %s",filename,strerror(errno));
         exit(1);
      }
      bytes_written += interval;
         
      one_sec_max = get_peak_value((short *)wavbuf,(short *)(wavbuf + interval));
      newtScaleSet(vu_1sec,one_sec_max);

      sprintf(labelstr_1sec,"%1.6f",((double )one_sec_max/ (double )SHRT_MAX));
      newtLabelSetText(label_1sec,labelstr_1sec);
      sprintf(labelstr_length,"%7ld",
         bytes_written / wav_header.nAvgBytesPerSec);
      newtLabelSetText(label_length,labelstr_length);
      if (one_sec_max > total_max)
      {
         total_max = one_sec_max;
         if (total_max > (SHRT_MAX - 1))
         {
            sprintf(labelstr_total,"CLIPPED");
         }
         else
         {
            sprintf(labelstr_total,"%1.6f",
               ((float )total_max/ (float )SHRT_MAX));
         }
         newtLabelSetText(label_total,labelstr_total);
         newtScaleSet(vu_total,total_max);
      }
      rf_result = newtRunForm(mainwaveform);
      
      if (rf_result == NULL)
      {
         running = 0;
      }

      newtRefresh();
   }
   newtFormDestroy(mainwaveform);
   newtFinished();         

   /* set the wav file header to have the right length */
   wav_header_on_disk = (WAVE_HEADER *)mmap(NULL,sizeof(WAVE_HEADER),
      PROT_READ | PROT_WRITE,MAP_SHARED,wavfd,0);
   if (wav_header_on_disk == MAP_FAILED)
   {
      printf("Error: mmap() %s\n",strerror(errno));
      printf("The .wav will be OK but you may have to fix the length\n");
      exit(1);
   }
   wav_header_on_disk->nDataBytes = bytes_written;
   munmap(wav_header_on_disk,sizeof(WAVE_HEADER));

   free(wavbuf);
   close(wavfd);               
   close(dspfd);


      
   return;
}

/**

Opens up /dev/dsp to read audio data 

presently forces 16bit stereo 44100 hz (what else would we want?).

Returns the fd to thedsp, and sets the block size in blk_size.

**/

int open_dsp_read(char *dsp_filename,size_t *blk_size)
{
   int dspfd;
   int status;
   int arg;

   /** open the sound device and set it up **/
   dspfd = open(dsp_filename,O_RDONLY);
   if (dspfd < 0) 
   {
      printf("Error: open(%s) %s\n", dsp_filename,strerror(errno));
      return(-1);
   }
   arg = BITS_PER_SAMPLE;
   status = ioctl(dspfd, SOUND_PCM_WRITE_BITS, &arg);

   if (status == -1)
   {
      perror("SOUND_PCM_WRITE_BITS ioctl failed");
      return(-1);
   }
   if (arg != BITS_PER_SAMPLE)
   {
      perror("unable to set sample size");
      return(-1);
   }

   arg = NUMBER_OF_CHANNELS;  /* mono or stereo */
   status = ioctl(dspfd, SOUND_PCM_WRITE_CHANNELS, &arg);
   if (status == -1)
   {
      perror("SOUND_PCM_WRITE_CHANNELS ioctl failed");
      return(-1);
   }
   if (arg != NUMBER_OF_CHANNELS)
   {
      perror("unable to set number of channels");
      return(-1);
   }

   arg = SAMPLES_PER_SECOND;      /* sampling rate */
   status = ioctl(dspfd, SOUND_PCM_WRITE_RATE, &arg);
   if (status == -1)
   {
      perror("SOUND_PCM_WRITE_RATE ioctl failed");
      return(-1);
   }
   status = ioctl(dspfd, SNDCTL_DSP_GETBLKSIZE,&arg);
   /* printf("status %d, blksize %d\n",status,arg); */
   *blk_size = arg;
   return(dspfd);
}

void usage(char *progname)
{
   printf("%s [ -d device ] file.wav\n", progname);
   printf(" Version %s duncan@zog.net.au (c) 1999\n",VERSION);
}

